File: 1466. Reorder Routes to Make All Paths Lead to the City Zero/Solution.java
class Solution {
    public int minReorder(int n, int[][] connections) {
        UFDS uf = new UFDS(n);
        Deque<int[]> d = new ArrayDeque<>();
        for (int[] c : connections) {
            int a = uf.find(c[0]), b = uf.find(c[1]);
            if (a == 0 || b == 0) {
                uf.union(a, b);
            } else {
                d.offerLast(c);
            }
        }
        while (!d.isEmpty()) {
            int[] c = d.pollLast();
            int a = uf.find(c[0]), b = uf.find(c[1]);
            if (a == 0 || b == 0) {
                uf.union(a, b);
            } else {
                d.offerFirst(c);
            }
        }
        return uf.count;
    }
}

class UFDS {
    int[] parent;
    int count;

    public UFDS(int n) {
        parent = new int[n];
        count = 0;
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) return;
        int a = find(x), b = find(y);
        if (a == 0) {
            count++;
            parent[b] = a;
        } else {
            parent[a] = b;
        }
    }
}

File: 1466. Reorder Routes to Make All Paths Lead to the City Zero/Solution.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        s1, s2 = [], []
        uf = UFDS(n)
        for x, y in connections:
            a, b = uf.find(x), uf.find(y)
            if a == 0 or b == 0:
                uf.union(a, b)
            else:
                s1.append([x, y])
        while s1 or s2:
            while s1:
                x, y = s1.pop()
                a, b = a, b = uf.find(x), uf.find(y)
                if a == 0 or b == 0:
                    uf.union(a, b)
                else:
                    s2.append([x, y])
            while s2:
                x, y = s2.pop()
                a, b = a, b = uf.find(x), uf.find(y)
                if a == 0 or b == 0:
                    uf.union(a, b)
                else:
                    s1.append([x, y])
        return uf.count

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.count = 0

    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def union(self, x, y):
        if self.same(x, y): return
        a, b = self.find(x), self.find(y)
        if a == 0:
            self.parent[b] = a
            self.count += 1
        else:
            self.parent[a] = b